go/store/datas/puller.go

```diff
@@ -166,37 +167,105 @@ func NewPuller(ctx context.Context, tempDir string, chunksPerTF int, srcDB, sink
 		return nil, err
 	}
 
-	return &Puller{
+	logFilePath := filepath.Join(tempDir, "push.log")
+	f, err := os.OpenFile(logFilePath, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, os.ModePerm)
+
+	var pushLogger *log.Logger
+	if err == nil {
+		pushLogger = log.New(f, "", log.Lmicroseconds)
+	}
+
+	p := &Puller{
 		fmt:           srcDB.Format(),
 		srcDB:         srcDB,
 		srcChunkStore: srcChunkStore,
-		sinkDB:        sinkDB,
+		sinkDBCS:      sinkDBCS,
 		rootChunkHash: rootChunkHash,
 		downloaded:    hash.HashSet{},
+		tablefileSema: semaphore.NewWeighted(outstandingTableFiles),
 		tempDir:       tempDir,
 		wr:            wr,
 		chunksPerTF:   chunksPerTF,
 		eventCh:       eventCh,
-	}, nil
+		pushLog:       pushLogger,
+	}
+
+	lcs, ok := sinkDBCS.(interface {
+		SetLogFunc(func(string, ...interface{}))
+	})
+
+	if ok {
+		lcs.SetLogFunc(p.log)
+	}
+
+	return p, nil
 }
 
-func (p *Puller) processCompletedTables(ctx context.Context, ae *atomicerr.AtomicError, completedTables <-chan FilledWriters) {
-	type tempTblFile struct {
-		id          string
-		path        string
-		numChunks   int
-		contentLen  uint64
-		contentHash []byte
+func (p *Puller) log(fmt string, args ...interface{}) {
+	if p.pushLog != nil {
+		if len(args) > 0 {
```

Comment:
You don't need to check if the slice is empty for vararg expansion like this, works fine

https://play.golang.org/p/pB9g1NB6jll